rust_src/build.rs

```diff
@@ -0,0 +1,206 @@
+                    has_lisp_fns = true;
+                } else if line.starts_with("#[no_mangle]") {
+                    parse_state = ParseState::NoMangleFn;
+                } else if line.starts_with("/*") {
+                    if !line.ends_with("*/") {
+                        parse_state = ParseState::IgnoreComments;
+                    }
+                } else if line.starts_with("include!(concat!(env!(\"OUT_DIR\"),") {
+                    has_include = true;
+                }
+            }
+
+            ParseState::IgnoreComments => if line.starts_with("*/") || line.ends_with("*/") {
+                parse_state = ParseState::Looking;
+            },
+
+            ParseState::LispFn(name) => {
+                if line.starts_with("pub") || line.starts_with("fn") {
+                    if let Some(func) = name.or_else(|| get_function_name(&line)) {
+                        write!(out_file, "pub use {}::F{};\n", modname, func)?;
+                        exported.push(func);
+                    }
+                }
+
+                parse_state = ParseState::Looking;
+            }
+
+            // export public #[no_mangle] functions
+            ParseState::NoMangleFn => {
+                if line.starts_with("pub") {
+                    if let Some(func) = get_function_name(&line) {
+                        write!(out_file, "pub use {}::{};\n", modname, func)?;
+                    }
+                }
+
+                parse_state = ParseState::Looking;
+            }
+        };
+    }
+
+    if has_lisp_fns {
+        let path =
+            PathBuf::from(env::var("OUT_DIR").unwrap()).join([modname, "_exports.rs"].concat());
+        let mut exports_file = File::create(path)?;
+
+        write!(
+            exports_file,
+            "export_lisp_fns! {{ {} }}",
+            exported.join(", ")
+        )?;
+
+        if !has_include {
+            panic!(
+                [
+                    modname,
+                    ".rs is missing the required include for lisp_fn exports"
+                ].concat()
+            );
+        }
+    }
+
+    Ok(has_lisp_fns)
+}
+
+fn ignore(path: &str) -> bool {
+    path == "" || path.starts_with('.') || path == "lib.rs"
+}
+
+fn generate_c_exports() -> Result<(), io::Error> {
+    let out_path = PathBuf::from(env::var("OUT_DIR").unwrap()).join("c_exports.rs");
+    let mut out_file = File::create(out_path)?;
+
+    let mut modules: Vec<String> = Vec::new();
+
+    let in_path = PathBuf::from(env::var("CARGO_MANIFEST_DIR").unwrap()).join("src");
+    for entry in fs::read_dir(in_path)? {
+        let entry = entry?;
+        let mut mod_path = entry.path();
+
+        if ignore(mod_path.file_name().unwrap().to_str().unwrap()) {
+            continue;
+        }
+
+        let mut name: Option<String> = None;
+
+        if mod_path.is_dir() {
+            name = Some(
+                mod_path
+                    .file_name()
+                    .and_then(|s| s.to_str())
```

Comment:
You've already `unwrap()`ped the `mod_path.file_name()` once above, so handling failure differently here seems unnecessary.

Reply:
I think this was the first code. I will give it a once over.